purview/utils/teiid.py

```diff
@@ -56,35 +58,44 @@ def get_connection(self, db_name):
         self._connections[db_name] = conn
         return conn
 
-    def query(self, sql, db='public', retry=3):
+    def query(self, sql, db='public', retry=None):
         """
         Send the SQL query to Teiid and return the rows as a list.
 
         :param str sql: the SQL query to send to the database
         :kwarg str db: the database name to query on
-        :kwarg int retry: the  number of times to retry a failed query
+        :kwarg int retry: the  number of times to retry a failed query. If this
+        is not set, then the TEIID query will be repeated until it is successful.
         :return: a list of rows from Teiid. Each row is a dictionary
         with the column headers as the keys.
         :rtype: list
         """
         con = self.get_connection(db)
         cursor = con.cursor()
-        if retry < 1:
+        if retry is not None and retry < 1:
             raise ValueError('The retry keyword must contain a value greater than 0')
 
         log.debug('Querying Teiid DB "{0}" with SQL:\n{1}'.format(db, sql))
 
+        fifteen_mins = 15 * 60
+        backoff = 30
         attempts = 0
         while True:
             attempts += 1
             try:
                 cursor.execute(sql)
                 break
             except psycopg2.extensions.QueryCanceledError:
-                if attempts < retry:
-                    log.warning('Teiid query failed')
-                else:
+                if retry is not None and attempts > retry:
                     raise
+                else:
+                    # Double the backoff time
+                    backoff = backoff * 2
+                    # Max out the backoff time to 15 minutes
+                    backoff = min((backoff, fifteen_mins))
```

Comment:
Instead of doing this, I would suggest checking `if backoff < fifteen_mins:` before doing `'backoff*2'` because the current approach might end up taking a lot of memory in case teiid goes down for a day or so

Reply:
Good point. I fixed that.